### PR TITLE
Fix force wait for keys

### LINKED ISSUE #1404: Client does not time out when looking for network on its own port
If the client binds its own qp2p port to the default (12000) and it's looking for a network on the default port (12000), it seems like the client does not time out while looking for the network (bootstrapping).

I've tried this with the `QuicP2pConfig.idle_timeout_msec` set to 5000 (5 seconds). Providing a hard coded contact with a different port (e.g. 12001), the client 'successfully' times out.

### PR DESCRIPTION
**Pull request by joshuef**
Resolves #1404

### CODE REVIEW

File: src/connection_manager.rs [hunk 1]
```diff
@@ -160,19 +173,33 @@ impl ConnectionManager {
     pub async fn send_transfer_validation(
         msg: &Message,
         sender: Sender<Result<TransferValidated, Error>>,
-        endpoint: Endpoint,
-        elders: Vec<SocketAddr>,
-        pending_transfers: PendingTransferValidations,
+        session: &Session,
+        // endpoint: Endpoint,
+        // elders: Vec<SocketAddr>,
+        // pending_transfers: PendingTransferValidations,
```
Comment (Yoga07) [inline src/connection_manager.rs:179]: Can remove these.

**Review by Yoga07 — DISMISSED** [review]

File: Cargo.toml [hunk 2]
```diff
@@ -29,7 +29,7 @@ rmp-serde = "~0.15.4"
 self_encryption = "~0.20.0"
 serde_json = "1.0.62"
 sn_data_types = "~0.15.0"
-sn_messaging = "~6.0.0"
+sn_messaging = { git = "https://github.com/oetyng/sn_messaging", branch = "split-section-funds" }
```
[the remark this answers is not in the document]
Comment (joshuef) [reply Cargo.toml:32]: ahhh, aye right. _this_ was the blocker.

File: src/client/transfer_actor/balance_management.rs [hunk 3]
```diff
@@ -140,17 +136,21 @@ impl Client {
...
-            .await
-            .transfer(amount, to, "asdf".to_string())?
-            .ok_or(Error::NoTransferGenerated)?;
+        {
+            info!(
+                "Our actor balance at send: {:?}",
+                self.transfer_actor.lock().await.balance()
+            );
+        }
+        let initiated;
```
Comment (lionel1704) [inline src/client/transfer_actor/balance_management.rs:145]: Can't this be
```suggestion
        let initiated = { ... };
```

File: src/connection_manager.rs [hunk 4]
```diff
@@ -57,47 +57,61 @@ pub struct ConnectionManager {}
...
-                Err(e) => Err(e.into()),
-            }
-        } else {
-            trace!("incoming_messages.next().await was None");
-            Err(Error::NotBootstrapped)
+        // bootstrap is not complete until we have pk set...
+        while !we_have_keyset {
+            use tokio::time::{delay_for, Duration};
+            delay_for(Duration::from_millis(500)).await;
+            we_have_keyset = session.section_key_set.lock().await.is_some();
```
Comment (lionel1704) [inline src/connection_manager.rs:91]: I still think we can prevent this looping here

File: src/connection_manager.rs [hunk 5]
```diff
@@ -57,47 +57,61 @@ pub struct ConnectionManager {}
...
+        // elders: Vec<SocketAddr>,
     ) -> Result<(), Error> {
         let msg_id = msg.id();
+        let endpoint = session.endpoint()?.clone();
+
+        let elders: Vec<SocketAddr>;
+
+        {
+            elders = session.elders.lock().await.iter().cloned().collect();
+        }
```
Comment (lionel1704) [inline src/connection_manager.rs:111]:
```suggestion
        let elders = {
             session.elders.lock().await.iter().cloned().collect();
        }
```
Perhaps even without the braces. It should remain the same...

File: src/connection_manager.rs [hunk 6]
```diff
@@ -160,19 +173,33 @@ impl ConnectionManager {
...
             "Sending transfer validation command {:?} w/ id: {:?}",
             msg,
             msg.id()
         );
+        let endpoint = session.endpoint()?.clone();
+        let elders: Vec<SocketAddr>;
+
+        {
+            elders = session.elders.lock().await.iter().cloned().collect();
+        }
```
Comment (lionel1704) [inline src/connection_manager.rs:191]: Here too

File: src/connection_manager.rs [hunk 7]
```diff
@@ -201,12 +228,16 @@ impl ConnectionManager {
...
-        elders: Vec<SocketAddr>,
-        pending_queries: PendingQueryResponses,
-    ) -> Result<QueryResponse, Error> {
+    pub async fn send_query(msg: &Message, session: &Session) -> Result<QueryResponse, Error> {
+        let endpoint = session.endpoint()?.clone();
+        let elders: Vec<SocketAddr>;
+
+        {
+            elders = session.elders.lock().await.iter().cloned().collect();
+        }
```
Comment (lionel1704) [inline src/connection_manager.rs:237]: One more.

File: src/connection_manager.rs [hunk 8]
```diff
@@ -422,44 +453,65 @@ impl ConnectionManager {
...
+    ) -> Result<Session, Error> {
+        if session.is_connecting_to_new_elders {
+            debug!("Already attempting elder connections, dropping get_section call until that is complete.");
+            return Ok(session)
+        }
+        let elders: Vec<SocketAddr>;
+        
+        {
+            elders = session.elders.lock().await.iter().cloned().collect();
+        }
```
Comment (lionel1704) [inline src/connection_manager.rs:470]: One more

File: src/connection_manager.rs [hunk 9]
```diff
@@ -422,44 +453,65 @@ impl ConnectionManager {
...
         let mut tasks = Vec::default();
 
         let endpoint = session.endpoint()?;
         let msg = session.bootstrap_cmd().await?;
-        let peers = session.elders.clone();
+
+        let peers;
+        {
+            peers = session.elders.lock().await.clone();
+        }
```
Comment (lionel1704) [inline src/connection_manager.rs:513]: Here too

File: src/connection_manager.rs [hunk 10]
```diff
@@ -592,18 +667,44 @@ impl ConnectionManager {
         mut session: Session,
         info: &SectionInfo,
     ) -> Result<Session, Error> {
+        let original_elders ;
+
+        {
+            original_elders = session.elders.lock().await.clone();
+        }
```
Comment (lionel1704) [inline src/connection_manager.rs:674]: Here too

File: src/client/transfer_actor/balance_management.rs (hunk 3 above)
Comment (joshuef) [reply src/client/transfer_actor/balance_management.rs:145]: why though?

File: src/connection_manager.rs (hunk 4 above)
Comment (joshuef) [reply src/connection_manager.rs]: We need to, as we cannot guarantee a first repsonse is for our section. So we want to hit the standard looping code.

We dont _have_ to have a delay here... but seems better than just hammering things.

**Review by lionel1704 — DISMISSED** [review]

**Review by lionel1704 — APPROVED** [review]

**Pull request merged**